onnxruntime/core/quantization/quantization.h

```diff
@@ -0,0 +1,195 @@
+  }
+
+  float scale;
+  T zero_point;
+};
+
+// Returns quantization params from scale and zero point Tensor pointers.
+// Caller is responsible for assuming that both Tensor pointers are of valid
+// shape and type.
+template <typename T>
+Params<T> GetTensorQuantizationParams(const Tensor* scale_tensor,
+                                      const Tensor* zero_point_tensor) {
+  ORT_STATIC_ASSERT_QUANTIZATION_TYPES(T)
+  return Params<T>(
+      *(scale_tensor->template Data<float>()),
+      *(zero_point_tensor->template Data<T>()));
+}
+
+// Quantizes a given float value with provided quantization params.
+template <typename T>
+T Quantize(const float value, const Params<T>& params) {
+  T quant_value;
+  MlasQuantizeLinear(&value,
+                     &quant_value,
+                     /*N=*/1,
+                     params.scale,
+                     params.zero_point);
+  return quant_value;
+}
+
+// Quantizes a list of float values with provided quantization params.
+template <typename T>
+void Quantize(const float* data,
+              T* output,
+              const Params<T>& params,
+              size_t size) {
+  MlasQuantizeLinear(data, output, /*N=*/size, params.scale, params.zero_point);
+}
+
+// Quantizes a vector of float values with provided quantization params.
+template <typename T>
+void Quantize(const std::vector<float>& data,
+              std::vector<T>& output,
+              const Params<T>& params) {
+  ORT_ENFORCE(data.size() == output.size(),
+              "Input and output data must have the same length.");
+  Quantize(data.data(), output.data(), params, data.size());
+}
+
+// Calculates and returns linear quantization params for a given float buffer.
+// Output buffer is quantized with calculated params. The option to force
+// symmetric quantization for signed values (zero point is forced at 0) is
+// provided through the force_symmetric bool in-param.
+template <typename T>
+Params<T> QuantizeLinear(const float* data,
+                         T* output,
+                         size_t size,
+                         bool force_symmetric = false) {
+  Params<T> params;
+
+  T T_min = std::numeric_limits<T>::min();
+  T T_max = std::numeric_limits<T>::max();
+  // NOTE: ORT currently clamps signed quantization values to -127,127 instead
+  //       of -128/127. This is done to ensure that forced symmetric
+  //       quantization results in a zero point of exactly 0 for signed 8 bit
+  //       ints.
+  // TODO(kreeger): Consider adjusting this clamping to enable more precision
+  //                for signed 8 bit ints.
+  //                See quant_utils.py - get_qmin_qmax_for_qType() for impl.
+  if constexpr (std::is_same<T, int8_t>::value) {
+    T_min = -127;
+  }
+
+  float min = std::numeric_limits<float>::max();
+  float max = std::numeric_limits<float>::min();
+  for (size_t i = 0; i < size; ++i) {
+    min = std::min(min, data[i]);
+    max = std::max(max, data[i]);
+  }
```

Comment:
use std::minmax_element?

Reply:
Thought about it - but `std::minmax_element` requires an iterator - and this signature operates on a type pointer:

https://en.cppreference.com/w/cpp/algorithm/minmax_element

Reply:
pointer is a type of iterator. minmax_element(data, data+size) should work.

Reply:
Cool - done.